runtime/src/bank.rs

```diff
@@ -4682,16 +4741,25 @@ impl Bank {
         ProgramAccountLoadResult::InvalidAccountData
     }
 
-    pub fn load_program(&self, pubkey: &Pubkey, reload: bool) -> Arc<LoadedProgram> {
+    pub fn load_program(
+        &self,
+        pubkey: &Pubkey,
+        reload: bool,
+        recompile: Option<Arc<LoadedProgram>>,
+    ) -> Arc<LoadedProgram> {
         let loaded_programs_cache = self.loaded_programs_cache.read().unwrap();
-        let environments = loaded_programs_cache.get_environments_for_epoch(self.epoch);
-
+        let effective_epoch = if recompile.is_some() {
+            loaded_programs_cache.latest_root_epoch.saturating_add(1)
```

Comment:
I think the recompilation continues in the new epoch if there are programs in the queue. Is that correct? That would cause the effective epoch to be the next epoch of the new epoch.

Reply:
`latest_root_epoch` only is advanced to the new epoch when the recompilation phase ends (first rerooting in new epoch).
So it will point to the new epoch (not the one after) even if `self.epoch` is `latest_root_epoch + 1` here already.